File: flow/src/main/java/flow/FlowModelUser.java
package flow;

import java.util.Collection;
import java.util.Map;
import java.util.Set;

import androidx.annotation.NonNull;

public interface FlowModelUser {
  class Relations {
    @NonNull
    private final Map<Class, String> relations;

    public Relations(@NonNull Map<Class, String> relations) {
      this.relations = relations;
    }

    Map<Class, String> getRelations() {
      return relations;
    }

    Set<Class> getScopes() {
      return relations.keySet();
    }

    Collection<String> getTags() {
      return relations.values();
    }
  }

  @NonNull
  Relations getRelations();
}

File: flow/src/main/java/flow/FlowModelUser.kt
package flow

interface FlowModelUser {
  val relations: Relations

  class Relations(
    private val relations: Map<Class<*>, String>
  ) {
    internal val scopes: Set<Class<*>>
      get() = relations.keys

    internal fun tag(scope: FlowModelScope): String? =
      relations.get(scope::class.java)
  }
}